src/modules/iotjs_module_mqtt.c

```diff
@@ -0,0 +1,677 @@
+
+  uint16_t package_id = ack->packet_identifier;
+
+  // The callback takes the packet identifier as parameter.
+  iotjs_jargs_t args = iotjs_jargs_create(1);
+  iotjs_jargs_append_number(&args, package_id);
+
+  jerry_value_t fn = iotjs_jval_get_property(jsref, name);
+  iotjs_make_callback(fn, jsref, &args);
+  jerry_release_value(fn);
+  iotjs_jargs_destroy(&args);
+}
+
+
+JS_FUNCTION(MqttInit) {
+  DJS_CHECK_THIS();
+
+  jerry_value_t jtls = JS_GET_THIS();
+
+  iotjs_mqtt_client_t *mqtt_client = iotjs_mqtt_client_create(jtls);
+
+  IOTJS_UNUSED(mqtt_client);
+
+  return jerry_create_undefined();
+}
+
+
+JS_FUNCTION(MqttConnect) {
+  DJS_CHECK_THIS();
+
+  DJS_CHECK_ARGS(1, object);
+
+  jerry_value_t joptions = JS_GET_ARG(0, object);
+
+  jerry_value_t jclient_id =
+      iotjs_jval_get_property(joptions, IOTJS_MAGIC_STRING_CLIENTID);
+  jerry_value_t jusername =
+      iotjs_jval_get_property(joptions, IOTJS_MAGIC_STRING_USERNAME);
+  jerry_value_t jpassword =
+      iotjs_jval_get_property(joptions, IOTJS_MAGIC_STRING_PASSWORD);
+  jerry_value_t jkeepalive =
+      iotjs_jval_get_property(joptions, IOTJS_MAGIC_STRING_KEEPALIVE);
+
+  uint8_t connect_flags = 0;
+  uint8_t keep_alive_msb = 0;
+  uint8_t keep_alive_lsb = 10;
+  connect_flags |= MQTT_FLAG_CLEANSESSION;
+
+  uint8_t header_byte = 0;
+  header_byte |= (CONNECT << 4);
+
+  if (!jerry_value_is_undefined(jusername)) {
+    connect_flags |= MQTT_FLAG_USERNAME;
+  }
+  if (!jerry_value_is_undefined(jpassword)) {
+    connect_flags |= MQTT_FLAG_PASSWORD;
+  }
+  if (!jerry_value_is_undefined(jkeepalive)) {
+    iotjs_mqtt_calculate_msb_lsb(jerry_get_number_value(jkeepalive),
+                                 &keep_alive_msb, &keep_alive_lsb);
+  }
+
+  iotjs_bufferwrap_t *client_id = iotjs_bufferwrap_from_jbuffer(jclient_id);
+  iotjs_mqtt_payload_t client_id_payload = iotjs_create_payload(client_id);
+
+  unsigned char variable_header_protocol[7];
+  variable_header_protocol[0] = 0;
+  variable_header_protocol[1] = 4;
+  variable_header_protocol[2] = 'M';
+  variable_header_protocol[3] = 'Q';
+  variable_header_protocol[4] = 'T';
+  variable_header_protocol[5] = 'T';
+  variable_header_protocol[6] = 4;
+
+  size_t variable_header_len = sizeof(variable_header_protocol) +
+                               sizeof(connect_flags) + sizeof(keep_alive_lsb) +
+                               sizeof(keep_alive_msb);
+
+  size_t payload_len = client_id_payload.size;
+  uint32_t remaining_length = payload_len + variable_header_len;
+  char *enc_remaining_length = iotjs_encode_remaining_length(remaining_length);
+  size_t full_len = sizeof(header_byte) + payload_len + variable_header_len +
+                    strlen(enc_remaining_length);
+
+  jerry_value_t jbuff = iotjs_bufferwrap_create_buffer(full_len);
+  iotjs_bufferwrap_t *buffer_wrap = iotjs_bufferwrap_from_jbuffer(jbuff);
+
+  uint8_t *buff_ptr = (uint8_t *)buffer_wrap->buffer;
+
+  // Fixed header
+  *buff_ptr++ = header_byte;
```

Comment:
I think this is the typical case when you don't need comment. It is obvious what you do here.